### PR TITLE
Basic GUI and basic messages and conversations update

### PR DESCRIPTION
**Pull request by heyimcesar**
- Added .gitignore
- Modified compile scripts
- ClientUser.java implementing database check function
- Modified the client and server controllers with "check in database" functionality
- Created "Conversation panel"
- Created "Message panel"
- Created "Menu bar panel"
- Login and logout callbacks
- Created DatabaseUser class for wrap the basic user information and loading it when server starts
- Added display name to the User class
- Search in the database users with HashMap with key "name", search in O(1)
- Added "Add file" GUI functionality
- When user creates a conversation it updates automatically the Message Panel
- Responsive design of the panels
- Update chat messages button installed

### CODE REVIEW

File: src/codeu/chat/server/Controller.java [hunk 1]
```diff
@@ -147,6 +164,67 @@ public Conversation newConversation(Uuid id, String title, Uuid owner, Time crea
...
+        if(userFromModel != null){
+          return userFromModel;
+        }else{
+          System.out.println("User in model but not in database");
+        }
+      }
+    }
+    return null;
+    // Iterate in every DatabaseUser
+    /*for(DatabaseUser databaseUser : databaseUsersList){
```
Comment (raguirre) [inline src/codeu/chat/server/Controller.java]: Same as before, here's a bunch of commented code that we most likely don't need anymore and it should not be committed to the repository.

File: src/codeu/chat/server/Controller.java [hunk 2]
```diff
@@ -147,6 +164,67 @@ public Conversation newConversation(Uuid id, String title, Uuid owner, Time crea
     return conversation;
   }
 
+  // Search for a user in the database and create the User object in the return.
+  @Override
+  public User searchUserInDatabase(String username, String pswd){
+    DatabaseUser tempUser = databaseUsersMap.get(username);
+    //return (tempUser != null) ? (tempUser.pswd.equals(pswd)) ? model.userById().first(tempUser.id) : null : null;
```
Comment (raguirre) [inline src/codeu/chat/server/Controller.java]: It's okay to comment out parts of your code temporarily when you don't know if you still need it or not, but you should clean up your code before committing.

File: src/codeu/chat/common/LogoutCallback.java [hunk 3]
```diff
@@ -0,0 +1,4 @@
+package codeu.chat.common;
+public interface LogoutCallback{
```
Comment (raguirre) [inline src/codeu/chat/common/LogoutCallback.java:2]: I get the idea of these Callback classes but I don't think we need such a simple interface, especially when it has one single method and is implemented only once. Maybe we could simply add the method to the class that implements it and get rid of this interface. That way we don't end up with a lot of unnecessary classes.

**Review by raguirre — APPROVED** [review]
The code looks good to me.

However, I have a few nitpicks regarding the overall workflow.
The PR is named "Basic GUI and basic messages and conversations update", nonetheless, it also includes changes to the building scripts, gitignore, login/logout, and database/user stuff.

You shouldn't include all these things into a single PR, especially when they are so unrelated to one another. I don't see why you had to wait to finish GUI work to send a PR on building scripts, for example.

I would recommend to work on smaller, more atomic and more manageable commits, and once you finish a certain task (i.e. building scripts), you send out a PR consisting of that single task, that way it's easier for us to look at fewer changes more frequently, than having to review a 20 files cl in one sitting.

Also this way everyone has the latest stuff asap.

Just my two cents and would like to see what the others have to say.

File: src/codeu/chat/common/LogoutCallback.java (hunk 3 above)
Comment (heyimcesar) [reply src/codeu/chat/common/LogoutCallback.java:2]: Wee need to implement the callback in order to pass the function to the other panels, it is a Java standard to create callbacks using interfaces.

File: compile_and_run_server.sh [hunk 4]
```diff
@@ -0,0 +1,6 @@
+if [ ! -d "./bin/com/" ]; then
+  unzip -d ./bin ./third_party/gson-2.8.0.jar
+fi
+sh make.sh
+cd './bin'
+java codeu.chat.ServerMain "100.101" "ABABAB" "2007"
```
Comment (wilph) [inline compile_and_run_server.sh]: What are these magic values?  (100.101, ababab, 2007)

File: compile_and_run_chat.sh [hunk 5]
```diff
@@ -1,6 +1,7 @@
 #!/bin/bash
```
Comment (wilph) [inline compile_and_run_chat.sh]: Same comment as three weeks ago -- you probably want a comment at the top saying what this script is for.  Maybe saying it needs to run in conjunction with compile_and_run_server.sh

File: src/codeu/chat/client/ClientUser.java [hunk 6]
```diff
@@ -150,4 +151,20 @@ public String showUserInfo(String uname) {
   public static void printUser(User user) {
     System.out.println(getUserInfoString(user));
   }
+
+  // Check for user in database
+  public User checkUserInDatabase(String username, String pswd){
+    return controller.searchUserInDatabase(username, pswd);
+  } 
+
+  // Sign in with existing User
```
Comment (wilph) [inline src/codeu/chat/client/ClientUser.java]: add comment about what's being returned, it's not obvious from the name.

File: src/codeu/chat/client/maingui/ConversationPanel.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        ConversationPanel.this.getAllConversations(listModel);
+      }
+    });
+
+    // User clicks Conversations Add button.
+    addButton.addActionListener(new ActionListener() {
+      @Override
+      public void actionPerformed(ActionEvent e) {
+        if (clientContext.user.hasCurrent()) {
+          final String s = (String) JOptionPane.showInputDialog(ConversationPanel.this, "Enter title:", "Add Conversation", JOptionPane.PLAIN_MESSAGE,null, null, "");
```
Comment (wilph) [inline src/codeu/chat/client/maingui/ConversationPanel.java]: You should probably wrap this line, i.e., start newlines after 80 or 100 char.

File: src/codeu/chat/client/maingui/ConversationPanel.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      }
+    });
+
+    getAllConversations(listModel);
+  }
+
+  // Populate ListModel - updates display objects.
+  private void getAllConversations(DefaultListModel<String> convDisplayList) {
+
+    clientContext.conversation.updateAllConversations(false);
```
Comment (wilph) [inline src/codeu/chat/client/maingui/ConversationPanel.java:116]: it's often helpful to put an inline comment with the variable name that the true/false is for, e.g.,

clientContext.conversation.updateAllConversations(false /* currentChanged */);

(Though, I still don't know what it's for given that variable name...)

File: src/codeu/chat/client/maingui/ConversationPanel.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    convDisplayList.clear();
+
+    for (final ConversationSummary conv : clientContext.conversation.getConversationSummaries()) {
+      convDisplayList.addElement(conv.title);
+    }
+  }
+
+  // Locate the Conversation object for a selected title string.
+  // index handles possible duplicate titles.
+  private ConversationSummary lookupByTitle(String title, int index) {
```
Comment (wilph) [inline src/codeu/chat/client/maingui/ConversationPanel.java:126]: Have you considered writing any unit tests?  This is a very easily unit testable function.

File: src/codeu/chat/client/maingui/MainGui.java [hunk 10]
```diff
@@ -53,25 +57,79 @@ private void initialize() {
...
   // Login request callback function 
   @Override
   public void onLoginRequest(String username, String pswd){
     
-    // Login function
-    JOptionPane.showMessageDialog(mainFrame,"User: " + username + "\nPswd: " + pswd);
+    // Search for user in the server's database
+    final User loginUser = clientContext.user.checkUserInDatabase(username, pswd);
+
+    if(loginUser != null){
```
Comment (wilph) [inline src/codeu/chat/client/maingui/MainGui.java]: pick a consistent style wrt to spaces before/after parenthesis.

Elsewhere you've written this like:
if (loginUser != null) {

File: src/codeu/chat/client/maingui/MainGui.java [hunk 11]
```diff
@@ -53,25 +57,79 @@ private void initialize() {
...
+
+          // User successfully signed in
+          JOptionPane.showMessageDialog(mainFrame,"Hello " + clientContext.user.getCurrent().display_name);
+          mainViewPanel.removeAll();
+          loadChatPanels();
+        }
+      }else{
+
+        // User is already logged
+        JOptionPane.showMessageDialog(mainFrame, "User is already logged");
```
Comment (wilph) [inline src/codeu/chat/client/maingui/MainGui.java]: "logged in", not "logged"

File: src/codeu/chat/client/maingui/MainGui.java [hunk 12]
```diff
@@ -53,25 +57,79 @@ private void initialize() {
...
+      }
+    }else{
+
+      // User not found in database or incorrect password
```
Comment (wilph) [inline src/codeu/chat/client/maingui/MainGui.java]: Comment is pretty redundant with the error message, you could remove.

File: src/codeu/chat/client/maingui/MenuBarPanel.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    this.setBorder(BorderFactory.createMatteBorder(0,0,1,0,Color.BLACK));
+
+    final JPanel userInfoPanel = new JPanel(new GridBagLayout());
+    GridBagConstraints userInfoConstrains = new GridBagConstraints();
+
+    JLabel userDisplayNameLabel,userNameLabel;
+    if(clientContext.user.hasCurrent()){
+      userDisplayNameLabel = new JLabel(clientContext.user.getCurrent().display_name);
+      userNameLabel = new JLabel("@" + clientContext.user.getCurrent().name);
+      userDisplayNameLabel.setFont(new Font(userDisplayNameLabel.getFont().getName(),Font.BOLD,userDisplayNameLabel.getFont().getSize()));
```
Comment (wilph) [inline src/codeu/chat/client/maingui/MenuBarPanel.java]: should probably wrap these lines, they're a bit long

File: src/codeu/chat/common/DatabaseUser.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+import codeu.chat.util.Serializers;
+
+// DatabaseUser is used to parse the json object from the database with Gson.
+public class DatabaseUser{
+    public final String id;
+    public final String display_name;
+    public final String pswd;
+    public final String name;
+
+    public DatabaseUser(String id, String display_name, String pswd, String name){
```
Comment (wilph) [inline src/codeu/chat/common/DatabaseUser.java]: display_name, not displayName?

**Review by wilph — COMMENTED** [review]
Had to skim a bit, just did a quick pass.

Ruben already mentioned this, but this is a very large pull request -- better practice to make many smaller pull requests, each doing one thing, than doing many different things at once.

File: compile_and_run_server.sh (hunk 4 above)
Comment (heyimcesar) [reply compile_and_run_server.sh]: I grab the code from the sh run_server.sh that CodeU already had

File: src/codeu/chat/client/ClientUser.java [hunk 15]
```diff
@@ -150,4 +151,21 @@ public String showUserInfo(String uname) {
   public static void printUser(User user) {
     System.out.println(getUserInfoString(user));
   }
+
+  // Check for user in database
+  public User checkUserInDatabase(String username, String pswd){
+    // If user matches with the username and password retreive the User from the server model if don't return null
```
Comment (YoaliSB) [inline src/codeu/chat/client/ClientUser.java:157]: nit:
I would write: 
"If user matches with the username  and password, retrieve the user from the server model. If not, return null."

File: src/codeu/chat/client/maingui/ConversationPanel.java [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+    final JButton addButton = new JButton("Add");
+
+    updateButton.setAlignmentX(Component.LEFT_ALIGNMENT);
+    buttonPanel.add(updateButton);
+    buttonPanel.add(addButton);
+
+    this.add(titlePanel, BorderLayout.NORTH);
+    this.add(listShowPanel, BorderLayout.CENTER);
+    this.add(buttonPanel, BorderLayout.SOUTH);
+
```
Comment (YoaliSB) [inline src/codeu/chat/client/maingui/ConversationPanel.java:67]: This is the first time I see someone order the UI elements in such an easy-to-understand and organized way. Nice.

File: src/codeu/chat/client/maingui/ConversationPanel.java (hunk 8 above)
Comment (YoaliSB) [reply src/codeu/chat/client/maingui/ConversationPanel.java:116]: I've never seen anyone do this, but it is indeed really helpful! I'll start doing this :)

**Review by YoaliSB — APPROVED** [review]

**Pull request merged**